### DISCUSSION TITLE
Cors setting for lambda handler

### DISCUSSION THREAD

**Discussion by perfectbase** [trpc/trpc#2371] [Q&A]
I'm using tRPC with serverless framework to deploy to ApiGateway+Lambda.

I want to find a way to add custom headers to the API response.

The closest answer I got was [this one](https://github.com/trpc/trpc/discussions/1985#discussioncomment-2947243):
```typescript
export const createContext = async ({ req, res }: CreateExpressContextOptions) => {
  return { req, res };
};

...

.query('...', {
  ...,
  resolve: async ({ ctx }) => {
    ctx.res.header('x-custom-header', 'hello123');
    ...
  },
}
```

But this is for the express handler.
I couldn't find a way to set the header for a lambda handler.
```typescript
export const createContext = ({
  event,
  context,
}: CreateAWSLambdaContextOptions<APIGatewayProxyEventV2>) => ({
  req: '???',
  res: '???',
});
```

Is there some function I don't know of that allows us to set headers inside the tRPC route?

The only workaround I could think of is by switching to the express handler and using `vendia/serverless-express`. Still haven't tried that, so not sure that would work.

I would love some guidance here.

Thanks in advance.

**Comment by KATT** [comment]
cc @Lilja

**Comment by bestickley** [comment]
Use the `responseMeta` key function within `awsLambdaRequestHandler` function. See how it's used for headers [here](https://github.dev/trpc/trpc/blob/next/packages/server/src/index.ts).

Example:

```ts
import { awsLambdaRequestHandler } from "@trpc/server/adapters/aws-lambda";
import { config } from "../../domain/config/app-config.js";
import { t, createContext } from "./trpc.js";
import { userRouter } from "./routers/user-router.js";

export const appRouter = t.router({
  user: userRouter,
});

// export type definition of API
export type AppRouter = typeof appRouter;

export const handler = awsLambdaRequestHandler({
  router: appRouter,
  createContext,
  responseMeta() {
    return {
      headers: {
        "Access-Control-Allow-Origin": getAccessControlAllowOrigin(
          config.stage
        ),
        "Access-Control-Allow-Methods": "GET,POST",
        "Access-Control-Allow-Headers": "authorization",
      },
    };
  },
});

function getAccessControlAllowOrigin(stage: string) {
  if (stage === "dev") {
    return "*";
  } else if (stage === "test") {
    return "https://test.example.com";
  } else if (stage === "prod") {
    return "https://example.com";
  } else {
    throw new Error(
      "Could not determine Access-Control-Allow-Origin from stage env var"
    );
  }
}

```

**Reply by zirkelc** [reply]
I tried to get CORS to work with the native API Gateway HTTP API but it just didn't work! Your example solved it. Do you know why it doesn't work out of the box?

**Reply by bestickley** [reply]
Hi @zirkelc, if you're proxying your Lambda with HTTP API GW (vs REST API GW), you're still responsible for setting the CORS headers on the request. If you're using an HTTP integration (vs Lambda) then you can configure within AWS what the response headers will be. But if you're using Lambda then those configurations won't have any effect.

Side note: keep in mind that if you're using HTTP API GW as your app API, you won't be able to protect it with WAF. You'll have to stick with REST API in order to use WAF. That's keeping me from switching to HTTP API GW from REST API GW.

**Reply by zirkelc** [reply]
Thank you @bestickley for the quick reply! I got it to work with HTTP API GW after lots of trial and error. First, I didn't know that I still have to send the `Access-Control-Allow-*` headers on the actual GET/POST requests. I thought enabling CORS on API level was sufficient. Second, I was using a catch-all route `*` route. I think this route intercepts the `OPTIONS` preflight requests as well. I changed it to two more specific routes `GET /{proxy+}`  and `POST /{proxy+}` and now it works.

**Reply by menno-lab** [reply]
Came across this thread when I kept running into CORS errors. I used your example to get most of the way there but I was still getting an error `Response to preflight request doesn't pass access control check: It does not have HTTP ok status.`

So I needed to adjust it slightly by adding a 200 status:

```tsx
export const handler = awsLambdaRequestHandler({
  router: appRouter,
  createContext,
  responseMeta() {
    return {
      status: 200,
      headers: {
        "Access-Control-Allow-Origin": getAccessControlAllowOrigin(
          config.stage
        ),
        "Access-Control-Allow-Methods": "GET,POST",
        "Access-Control-Allow-Headers": "authorization",
      },
    };
  },
});
```

Maybe this will help someone else who stumbles across this issue.

**Reply by zirkelc** [reply]
Interesting that you have to add `status: 200`. It works for me without any status:

```js
  responseMeta() {
    return {
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'OPTIONS,POST,GET',
        'Access-Control-Allow-Headers': '*',
      },
    };
  },
```

Are you using the REST API Gateway or the HTTP API Gateway?

**Reply by menno-lab** [reply]
I'm using HTTP API Gateway

**Comment by salza80** [comment]
> Thank you @bestickley for the quick reply! I got it to work with HTTP API GW after lots of trial and error. First, I didn't know that I still have to send the `Access-Control-Allow-*` headers on the actual GET/POST requests. I thought enabling CORS on API level was sufficient. Second, I was using a catch-all route `*` route. I think this route intercepts the `OPTIONS` preflight requests as well. I changed it to two more specific routes `GET /{proxy+}` and `POST /{proxy+}` and now it works.

This solved it for me! The problem was ANY /api/{proxy+}" with an authorizer. You can either split it out to GET, POST etc so it doesn't intefer with the OPTIONS preflight requests OR yo can add a OPTIONS /api/{proxy+} above it and handle these in another handler without authorization... and return the appropriate headers.
I found it documented here: https://aws.amazon.com/blogs/compute/configuring-cors-on-amazon-api-gateway-apis/. under "Authorization" heading.

**Comment by zirkelc** [comment]
I stumbled upon another CORS issue and wanted to share it here. I switched my Lambda function from API Gateway to the Lambda Function URL. This supports CORS as a configuration option similar to HTTP or REST API. However, the specified CORS headers (AllowOrigins, AllowMethods, ...) get added **automatically** by Lambda. That means, any additional headers specified via `responseMeta()` will get merged with the headers specified on the Lambda Function URL config. In my case, I had two entries for `Access-Control-Allow-Origin` on the POST response from TRPC and Chrome blocked this request. Removing `responseMeta()` solved this issue.
